### PR TITLE
qt5: fix build with apple-clang 15 + add package_type + bump deps + use version range for zlib

### PR DESCRIPTION
**Pull request by SpaceIm**
- fix build with apple-clang 15
- robust build on macOS, when dependencies like zlib are shared (because it's linked to moc executable, called during build process, and it may fail due to SIP)
- add package_type
- bump dependencies
- use `<host_version>` for wayland in build requirements
- use version range for zlib

closes https://github.com/conan-io/conan-center-index/issues/20075

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [ ] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by SpaceIm** [comment]
There is another issue to fix: https://github.com/conan-io/conan-center-index/issues/20075#issuecomment-1732561107

**Review by jwillikers — APPROVED** [review]

**Comment by SpaceIm** [comment]
Now it builds with apple-clang 15, but locally there is a SIP issue when zlib is shared... I though that I've fixed that in a previous PR, I hate Apple and their SIP.

```
moc /Users/spaceim/.conan2/p/qte60086837c50b/s/qt5/qtbase/src/corelib/animation/qabstractanimation.h
dyld[85681]: Library not loaded: @rpath/libz.1.dylib
  Referenced from: <CB1C7CDD-A1A2-321B-948B-CF92D328A0F1> /Users/spaceim/.conan2/p/b/qt2ad21b762eb26/b/build_folder/qtbase/bin/moc
  Reason: tried: '/Users/spaceim/.conan2/p/b/qt2ad21b762eb26/b/build_folder/qtbase/Frameworks/libz.1.dylib' (no such file), '/Users/spaceim/.conan2/p/b/qt2ad21b762eb26/b/build_folder/qtbase/Frameworks/libz.1.dylib' (no such file)
```

**Comment by SpaceIm** [comment]
> Now it builds with apple-clang 15, but locally there is a SIP issue when zlib is shared... I though that I've fixed that in a previous PR, I hate Apple and their SIP.
> 
> ```
> moc /Users/spaceim/.conan2/p/qte60086837c50b/s/qt5/qtbase/src/corelib/animation/qabstractanimation.h
> dyld[85681]: Library not loaded: @rpath/libz.1.dylib
>   Referenced from: <CB1C7CDD-A1A2-321B-948B-CF92D328A0F1> /Users/spaceim/.conan2/p/b/qt2ad21b762eb26/b/build_folder/qtbase/bin/moc
>   Reason: tried: '/Users/spaceim/.conan2/p/b/qt2ad21b762eb26/b/build_folder/qtbase/Frameworks/libz.1.dylib' (no such file), '/Users/spaceim/.conan2/p/b/qt2ad21b762eb26/b/build_folder/qtbase/Frameworks/libz.1.dylib' (no such file)
> ```

Fixed by latest commit https://github.com/conan-io/conan-center-index/pull/20093/commits/f9c91fa481d4ad295d46e94c2fe3d12cecaf12be

**Review by jwillikers — APPROVED** [review]

**Comment by SpaceIm** [comment]
v2 pipeline:

```
ERROR: Missing binary: fontconfig/2.14.2:be99654e60049e3c9c81374f129beeb1bf5469ab
ERROR: Missing binary: xkbcommon/1.5.0:bd827029f896951473e16bfc49e278400195e828
```

@RubenRBS I don't know if those one are available now.

**Review by MartinDelille — APPROVED** [review]

File: recipes/qt/5.x.x/conanfile.py [hunk 1]
```diff
@@ -801,6 +789,19 @@ def _getenvpath(var):
         if self._settings_build.os == "Linux" and self.settings.compiler == "clang":
             args += ['QMAKE_CXXFLAGS+="-ftemplate-depth=1024"']
 
+        if self._settings_build.os == "Macos":
+            # On macOS, SIP resets DYLD_LIBRARY_PATH injected by VirtualBuildEnv & VirtualRunEnv.
+            # Qt builds several executables (moc etc) which are called later on during build of
+            # libraries, and these executables link to several external dependencies in requirements().
+            # If these external libs are shared, moc calls fail because its dylib dependencies
```
Comment (RubenRBS) [inline recipes/qt/5.x.x/conanfile.py:796]: Where did the inspiration for this come from? Any extra insight would be appreciated :) but looks good to me

**Review by RubenRBS — COMMENTED** [review]
Thanks! Pinging @jcar87 as the it's a qt change

File: recipes/qt/5.x.x/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply recipes/qt/5.x.x/conanfile.py:796]: I have this error locally without this workaround: https://github.com/conan-io/conan-center-index/pull/20093#issuecomment-1732658091. Injection of VirtualRunEnv in build scope doesn't work on macOS due to SIP (at some point SIP resets DYLD_LIBRARY_PATH, I don't kno exactly when it happens during build, but it's a fact). Previous `BASH_ENV` workaround doesn't work either depending on terminal (zsh doesn't care for example). So if correct zlib version (matching the one installed by conan) is not installed on system, it fails.
The only workaround I've found is to add path to shared libs in LC_RPATH, SIP can't do anything against that.

Comment (SpaceIm) [reply recipes/qt/5.x.x/conanfile.py:796]: Actually, it's the same hack than the one in AutotoolsToolchain: https://github.com/conan-io/conan/blob/5f0dc4da2202069d20a59ccbbc17b179aaffed33/conan/tools/gnu/autotoolsdeps.py#L58-L60

**Comment by SpaceIm** [comment]
@RubenRBS is it fine for you?

**Comment by RubenRBS** [comment]
> I already had to handle a new version in this PR which was not there when I've opened this PR, it's a lot of work to maintain a PR when other more recent PRs are merged.

@SpaceIm I know it can get cumbersome, a high volume of PRs makes stuff like this happening more likely, so thanks again for taking the time to get it updated :)

**Pull request merged**